NPE when reading structs with a newly added subfield

In case of a schema change where a new field is added to a struct there will be a missing stream when reading from older files. We need to create a MissingFieldsStreamReader to return an empty block.
This was handled for the case where the field was in the output. The same should be added when its pruned.

```
Caused by: java.lang.NullPointerException: streamDescriptor is null
	at java.base/java.util.Objects.requireNonNull(Objects.java:246)
	at com.facebook.presto.orc.reader.StructSelectiveStreamReader$PruningStreamReader.<init>(StructSelectiveStreamReader.java:703)
	at com.facebook.presto.orc.reader.StructSelectiveStreamReader$PruningStreamReader.<init>(StructSelectiveStreamReader.java:690)
	at com.facebook.presto.orc.reader.StructSelectiveStreamReader.<init>(StructSelectiveStreamReader.java:191)
	at com.facebook.presto.orc.reader.SelectiveStreamReaders.createStreamReader(SelectiveStreamReaders.java:85)
	at com.facebook.presto.orc.reader.StructSelectiveStreamReader.<init>(StructSelectiveStreamReader.java:176)
	at com.facebook.presto.orc.reader.SelectiveStreamReaders.createStreamReader(SelectiveStreamReaders.java:85)
	at com.facebook.presto.orc.OrcSelectiveRecordReader.createStreamReaders(OrcSelectiveRecordReader.java:587)
	at com.facebook.presto.orc.OrcSelectiveRecordReader.<init>(OrcSelectiveRecordReader.java:179)
	at com.facebook.presto.orc.OrcReader.createSelectiveRecordReader(OrcReader.java:253)
	at com.facebook.presto.hive.orc.OrcSelectivePageSourceFactory.createOrcPageSource(OrcSelectivePageSourceFactory.java:371)
	... 17 more
```

cc: @mbasmanova  @sujay-jain  @jbroll  @oerling

--- a/presto-hive/src/test/java/com/facebook/presto/hive/TestHivePushdownFilterQueries.java
+++ b/presto-hive/src/test/java/com/facebook/presto/hive/TestHivePushdownFilterQueries.java
@@ -885,6 +885,8 @@ public class TestHivePushdownFilterQueries
         Path newDirectoryPath = getOnlyPath("test_struct_add_column").getParent();
         Files.move(oldFilePath, Paths.get(newDirectoryPath.toString(), "old_file"), ATOMIC_MOVE);
         assertQuery("SELECT * FROM test_struct_add_column", "SELECT (1, 2, 3) UNION ALL SELECT (1, 2, null)");
+        assertQuery("SELECT x.a FROM test_struct_add_column", "SELECT 1 UNION ALL SELECT 1");
+        assertQuery("SELECT count(*) FROM test_struct_add_column where x.c = 1", "SELECT 0");
     }
 
     @Test

--- a/presto-orc/src/main/java/com/facebook/presto/orc/reader/StructSelectiveStreamReader.java
+++ b/presto-orc/src/main/java/com/facebook/presto/orc/reader/StructSelectiveStreamReader.java
@@ -59,6 +59,7 @@ import static com.facebook.presto.orc.stream.MissingInputStreamSource.missingStr
 import static com.google.common.base.MoreObjects.toStringHelper;
 import static com.google.common.base.Preconditions.checkArgument;
 import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.base.Verify.verify;
 import static com.google.common.collect.ImmutableMap.toImmutableMap;
 import static com.google.common.collect.ImmutableSet.toImmutableSet;
 import static io.airlift.slice.SizeOf.sizeOf;
@@ -166,13 +167,17 @@ public class StructSelectiveStreamReader
                 boolean requiredField = requiredFields.map(names -> names.containsKey(fieldName)).orElse(outputRequired);
                 Optional<Type> fieldOutputType = Optional.ofNullable(nestedTypes.get(fieldName)).map(Field::getType);
 
-                if (requiredField || fieldsWithFilters.contains(fieldName)) {
-                    Map<Subfield, TupleDomainFilter> nestedFilters = filters.entrySet().stream()
-                            .filter(entry -> entry.getKey().getPath().size() > 0)
-                            .filter(entry -> ((Subfield.NestedField) entry.getKey().getPath().get(0)).getName().equalsIgnoreCase(fieldName))
-                            .collect(toImmutableMap(entry -> entry.getKey().tail(fieldName), Map.Entry::getValue));
-                    List<Subfield> nestedRequiredSubfields = requiredFields.map(names -> names.get(fieldName)).orElse(ImmutableList.of());
-                    if (nestedStream != null) {
+                if (nestedStream == null) {
+                    verify(fieldOutputType.isPresent(), "Missing output type for subfield " + fieldName);
+                    nestedReaders.put(fieldName, new MissingFieldStreamReader(fieldOutputType.get()));
+                }
+                else {
+                    if (requiredField || fieldsWithFilters.contains(fieldName)) {
+                        Map<Subfield, TupleDomainFilter> nestedFilters = filters.entrySet().stream()
+                                .filter(entry -> entry.getKey().getPath().size() > 0)
+                                .filter(entry -> ((Subfield.NestedField) entry.getKey().getPath().get(0)).getName().equalsIgnoreCase(fieldName))
+                                .collect(toImmutableMap(entry -> entry.getKey().tail(fieldName), Map.Entry::getValue));
+                        List<Subfield> nestedRequiredSubfields = requiredFields.map(names -> names.get(fieldName)).orElse(ImmutableList.of());
                         SelectiveStreamReader nestedReader = SelectiveStreamReaders.createStreamReader(
                                 nestedStream,
                                 nestedFilters,
@@ -184,11 +189,8 @@ public class StructSelectiveStreamReader
                         nestedReaders.put(fieldName, nestedReader);
                     }
                     else {
-                        nestedReaders.put(fieldName, new MissingFieldStreamReader(fieldOutputType.get()));
+                        nestedReaders.put(fieldName, new PruningStreamReader(nestedStream, fieldOutputType));
                     }
-                }
-                else {
-                    nestedReaders.put(fieldName, new PruningStreamReader(nestedStream, fieldOutputType));
                 }
             }
 

